torch/utils/_python_dispatch.py

```diff
@@ -165,3 +165,130 @@ def transform_subclass(t, callback):
     for attr in attrs:
         transformed_tensors_dict[attr] = callback(attr, getattr(t, attr))
     return type(t).__tensor_unflatten__(transformed_tensors_dict, ctx)
+
+def _correct_storage_aliasing(func, args, outs):
+    assert isinstance(func, torch._ops.OpOverload)
+    assert isinstance(args, tuple)
+    assert isinstance(outs, (list, tuple))
+    flat_outs, _ = torch.utils._pytree.tree_flatten(outs)
+    for x in flat_outs:
+        if isinstance(x, torch.Tensor):
+            assert hasattr(x, '__torch_dispatch__')
+
+    def alias_storage(arg, ret):
+        # This is hopefully a reasonable assert:
+        # subclasses that rely on this API for output aliasing
+        # should always return wrapper tensor subclasses for us to manually alias.
+        # in theory if a subclass that needs this API wants to sometimes return
+        # plain tensors, we could remove the assert and just not perform the aliasing,
+        # but it seems safer to learn more about this case first.
+        if is_traceable_wrapper_subclass(arg) or is_traceable_wrapper_subclass(ret):
+            assert type(arg) == type(ret), f"""Called {str(func)} with input of type {type(arg)}
+and output of type {type(ret)}. But expected types to match."""
+        # Need to run under no_dispatch, because we explicitly do **not**
+        # want our subclass to intercept the set_() call.
+        # instead, our subclass should directly have its storage swapped out.
+        with torch.utils._mode_utils.no_dispatch():
+            # directly calling this overload, and passing ret.shape, because we **explicitly**
+            # don't want to reset the sizes on ret, if the storage implies a size change.
+            # Why?
+            # The purpose of this API is *not* to change the size/strides of our output- we assume it's already correct.
+            # We just want to "fix up" the storage aliasing, without modifying or output's metadata.
+            # Example: out = inp.expand(inp.shape[0], inp.shape[0])
+            #     This requires swapping the storage of out to be the same as inp,
+            #     but we do *not* want it to change the sizes/strides that were compute for out.
+            meta_in_tls = torch._C._meta_in_tls_dispatch_include()
+            # TODO: make this a context manager
+            torch._C._set_meta_in_tls_dispatch_include(True)
+            try:
+                torch.ops.aten.set_.source_Storage_storage_offset(ret, arg.untyped_storage(), 0, ret.shape)
+            finally:
+                torch._C._set_meta_in_tls_dispatch_include(meta_in_tls)
```

Comment:
This is a bit sketchy: I want to slam the output storage onto the input (to guarantee correct aliasing), but:

(1) I use `no_dispatch()`, because I don't want the subclass to intercept my set_() call

(2) If I do that, the default behavior will be to dispatch to the cpu.cuda kernel for `set_()`. That's not actually what I want, since those two kernels don't support dynamic shapes (but the meta kernel does). That's why I swizzle the meta dispatch key here.

Reply:
darn, good catch

Reply:
Actually @albanD - I think Elias updated the `Meta` backend bit to be the highest priority backend bit ([code](https://github.com/pytorch/pytorch/blob/main/c10/core/DispatchKey.h#L49)).

This does mean though that backends that aren't part of `BackendComponent` might end up being higher priority:
- MkldnnCPU
- SparseCsrCPU/CUDA
- FPGA
- ORT
- Metal
- Vulkan


I wrote this function to do the meta swizzling the same way that it's done in fake tensor though, so I imagine that FakeTensor is equally broken with those backends. cc @eellison, what do you think?

For now, I can at least add a note here and reference it in fake tensor, so if/when we actually try to fix these backends later we'll know which code sites to audit.

Reply:
To be fair though, most of these backends feel like "special" backends that aren't going to work with `torch.compile` for a whole host of other reasons anyway:
- metal/vulkan meant for mobile
- ORT having its own custom backend anyway
- SparseCSR can/should eventually be re-written as a functioanlity key

Reply:
That is correct, the other ones will not work currently. Why are the backends not part of `BackendComponent` ? If we added them to the local exclude set, would it work then ? 

I think i remember there being something weird about exclude/include, so maybe that's not viable..

Reply:
>  Why are the backends not part of BackendComponent 

We have a pretty limited number of backend component slots (the "long term" solution is for more backends to use the out-of-tree PrivateUse1 registration. So we never ended up adding these extra backends here

> If we added them to the local exclude set, would it work then ? I think i remember there being something weird about exclude/include, so maybe that's not viable..

Actually, I think this should work. Although I'll leave it as not part of this PR

Reply:
Ho Elias is ahead of us haha!
Ignoring these non-generic backends sounds fine. cc @cpuhrsch FYI on the sparse one as it is the only one that may be used with compile.